### PR TITLE
Twingraph implementation

### PR DESCRIPTION
**Pull request by icreated**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -205,6 +205,8 @@ subprojects {
     val oktaSpringBootVersion = "2.1.6"
     implementation("com.okta.spring:okta-spring-boot-starter:${oktaSpringBootVersion}")
 
+    implementation("com.redislabs:jredisgraph:2.6.0-rc3")
```
Comment (vcarluer) [inline build.gradle.kts:208]: Maybe avoid rc release?

File: api/src/main/resources/application.yml [hunk 2]
```diff
@@ -48,6 +48,11 @@ server:
     context-path:
   forward-headers-strategy: native
 
+twin-graph:
```
Comment (vcarluer) [inline api/src/main/resources/application.yml:51]: move to csm.platform

File: twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphConfig.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    return RedisGraph(jedis())
+  }
+
+  @Bean
+  fun twingraphProperties(): TwinGraphProperties {
+    return TwinGraphProperties()
+  }
+}
+
+@ConfigurationProperties(prefix = "twin-graph")
```
Comment (vcarluer) [inline twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphConfig.kt:44]: csm.platform.

File: twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt [hunk 4]
```diff
@@ -21,7 +29,39 @@ class TwingraphServiceImpl : TwingraphApiService {
...
+      throw CsmResourceNotFoundException(
+          "TwinGraph empty with given ${twinGraphQuery.graphId} " +
+              "and version ${twinGraphQuery.version}")
+    }
+
+    val nodeList = mutableListOf<String>()
+    while (resultSet.hasNext()) {
+      val record: Record = resultSet.next()
+      val value = record.values()[0]
+      if (value is Node) {
```
Comment (vcarluer) [inline twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt:61]: And relationships?

**Review by vcarluer — CHANGES_REQUESTED** [review]
some changes

File: twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt [hunk 5]
```diff
@@ -21,7 +29,60 @@ class TwingraphServiceImpl : TwingraphApiService {
...
+      nextCursor = scanResult.cursor
+      matchingKeys.addAll(scanResult.result)
+    } while (!nextCursor.equals(SCAN_POINTER_START))
+
+    @Suppress("SpreadOperator") val count = jedis.del(*matchingKeys.toTypedArray())
+    logger.debug("$count keys are removed from Twingraph with prefix $graphId")
+  }
+
+  override fun query(organizationId: String, twinGraphQuery: TwinGraphQuery): String {
+    val redisGraphId = "${twinGraphQuery.graphId}:${twinGraphQuery.version}"
```
Comment (vcarluer) [inline twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt:46]: Make twinGraphQuery.version optional with default value = latest (fetch latest version from metadata)

File: twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt [hunk 6]
```diff
@@ -21,7 +30,75 @@ class TwingraphServiceImpl : TwingraphApiService {
...
+
+    @Suppress("SpreadOperator") val count = jedis.del(*matchingKeys.toTypedArray())
+    logger.debug("$count keys are removed from Twingraph with prefix $graphId")
+  }
+
+  override fun query(organizationId: String, twinGraphQuery: TwinGraphQuery): String {
+
+    if (twinGraphQuery.version == null) {
+      twinGraphQuery.version = jedis.hget("${twinGraphQuery.graphId}MetaData", "lastVersion")
+    }
```
Comment (jreynard-code) [inline twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt:50]: Handle case if lastVersion is not found or null. 
You can throw an exception (e.g. IllegalStateException)

File: twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt [hunk 7]
```diff
@@ -21,7 +30,75 @@ class TwingraphServiceImpl : TwingraphApiService {
...
+  override fun delete(graphId: String) {
+    val matchingKeys = mutableSetOf<String>()
+    var nextCursor = SCAN_POINTER_START
+    do {
+      var scanResult = jedis.scan(nextCursor, ScanParams().match("$graphId*"))
+      nextCursor = scanResult.cursor
+      matchingKeys.addAll(scanResult.result)
+    } while (!nextCursor.equals(SCAN_POINTER_START))
+
+    @Suppress("SpreadOperator") val count = jedis.del(*matchingKeys.toTypedArray())
```
Comment (jreynard-code) [inline twingraph/src/main/kotlin/com/cosmotech/twingraph/api/TwingraphServiceImpl.kt:42]: Could you please find a way to remove the `@Suppress("SpreadOperator")` ?

**Review by jreynard-code — CHANGES_REQUESTED** [review]
LGTM, just few little remarks ;)

**Pull request merged**